### ISSUE TITLE
Pull Replication issue: Handling Production SyncgateWays/ Couchbase clusters Swithovers.  

### ISSUE THREAD

**Issue by NitzDKoder** [couchbase/couchbase-lite-java-core#1651]
Hi Folks,

**Setup Details: -**
1.	We have 2 SG/CB-clusters (CB-cluster C1 & C2) and data is SG-replicated across clusters. 
2.	CBL has FQDN to discover the CB-clusters and connect. Both the clusters are identified by same FQDN.
3.	C1 is primary serving cluster, and if C1 fails over, C2 will be serving. When C1 is up, all CBL traffic will be routed to C1, and when C1 is down, all CBL traffic will be routed to C2.
                                
**Pre-Condition: -**
                                1. For given CB-lite client (CBL), Pull checkpoint document SeqNum of C2 is higher than C1.
                                2. Cluster C1 is down and C2 cluster is serving.

**Issues observed: -**
1. Observed Push & Pull connections are connected to different clusters. When C1 becomes active again after failure, new push connection will be routed to C1, but existing changes_feed connection to C2 is not closed by CBL.

2. Observed sometimes when failover happens from C2 to C1, CBL is updating checkpoint doc in C1 with seqNum of C2. If seqNum corresponds to checkpoint doc in C2 is higher than the seqNum corresponds to checkpoint doc in C1 (seqNum-C1 <seqNum-C2), then changes_feed from C1 won't give any notifications due to checkpoint document corruption.
                                (CBL receives changes_feed response with SeqNum-C2 from C2 -> CBL sends PUT request with checkpoint-SeqNum-C2 to C1 -> CBL receives 409 response from C1 -> CBL sends GET request to fetch checkpoint-doc to C1 -> CBL receives 200 OK response with checkpoint-SeqNum-C1 from C1 -> CBL sends PUT request with checkpoint-SeqNum-C2 to C1 -> CBL receives 200 OK response from C1).

3. On receiving 409 response for checkpoint doc update, CBL is not fetching checkpoint doc (it’s not re-starting replication).
                                                (CBL sends PUT request with checkpoint-SeqNum-C2 to C1 -> CBL receives 409 response from C1 -> CBL sends PUT request with checkpoint-SeqNum-C2 to C1 -> CBL receives 409 response from C1 ... repeat till replication restarts)

**Expectations: -**
                                i. CBL should be able to detect the CB-cluster failover/switchover, and shall be able to perform push & pull sync accordingly by initializing with appropriate checkpoints.
                                ii. CBL needs to make sure that Push & Pull checkpoint documents are updated with seqNum corresponds to its local cluster.
                                iii. On receiving 409 response for checkpoint doc update, CBL shall restarts it sync.
                                iv. On receiving 503 response or connection close for changes_feed, CBL shall restarts it sync.
                                v. CBL needs to make sure both the Push & Pull connection are going to same CB-cluster.

@hideki @adamcfraser @pasin @snej 
**Please share your thoughts to handle this use case in CBL or Server level.**

Thanks
Nithin

**Comment by ddash1** [comment]
@waynecarter @ssmotra 
Wayne, this is the issue I was talking about on both iOS and Android. Pls advise.

**Comment by snej** [comment]
>CBL should be able to detect the CB-cluster failover/switchover

It can't detect that directly. It only finds out when it tries to save a new checkpoint (which happens about every 5 seconds) and finds that the pre-existing checkpoint isn't there or has the wrong revID.

>v. CBL needs to make sure both the Push & Pull connection are going to same CB-cluster.

I don't know how it could tell. The switchover is presumably going on behind a reverse proxy or through a CDN switching DNS entries, so from the client side it doesn't look like anything's changed.

**Comment by snej** [comment]
>iii. On receiving 409 response for checkpoint doc update, CBL shall restarts it sync.

It doesn't currently do that. This needs more thought, but I think you're right that it should, since this implies the server may have been swapped out or restored from a backup or something. (The replication algorithm handles this situation when a replication starts, but I think it doesn't account for it happening in the middle of a replication.)

>iv. On receiving 503 response or connection close for changes_feed, CBL shall restarts it sync.

Dropped connections happen a lot, due to flaky networks or some middleware (like AWS) deciding that the socket seems idle. Resetting the replication every time this happened would be a huge performance problem for those situations, and might even mean that the replication never completes if the disconnect keeps happening.

Similarly, a 503 status just means "the server is unavailable", and it can happen for a lot of reasons. It would be a bad idea to reset replication whenever it happened.

**Comment by ddash1** [comment]
v. CBL needs to make sure both the Push & Pull connection are going to same CB-cluster.
I don't know how it could tell. The switchover is presumably going on behind a reverse proxy or through a CDN switching DNS entries, so from the client side it doesn't look like anything's changed.

Will this be handled better with the single (push/pull) replicator in 2.0?

**Comment by ddash1** [comment]
If we can address the 409 issue then we should be able to handle this for now.

**Comment by snej** [comment]
>Will this be handled better with the single (push/pull) replicator in 2.0?

Yes; a single Replication instance in 2.0 can do both push and pull. Also, replication runs over a single WebSocket connection, and if that socket closes unexpectedly the replicator will restart. The new replicator instance will compare checkpoints first thing. 

So this should resolve the problem discussed here: a failover to a different cluster would presumably terminate open WebSocket connections; the replicator will re-connect to the new cluster and then detect a mismatched checkpoint by resetting the replication to sequence 0.

**Comment by hideki** [comment]
Notes regarding `PUT /_local/{checkpointDocID}` and `409` (CONFLICT) cycle problem that is mentioned in observed item no. 3.

#### With the latest implementation of CBL Android 1.x (master branch), CBL should behave following:
1. `PUT /_local/{checkpointDocID}` failed with HTTP Error `409` (CONFLICT) because revision ID of checkpointDocID does not exist on the server. 
1. CBL sends `GET /_local/{checkpointDocID}`, and receives server's `_id`, `_rev` and `lastSequence` values.
1. CBL overwrites `lastSequence` value with local `lastSequence` value, and sends `PUT /_local/{checkpointDocID}` with updated data, this should be successful (Should receive HTTP OK).

#### Tested CBL by following steps to verify above expectation
1. Run push replication with continuous mode.
2. With using curl or equivarent tool, updates on server's checkpointDocID's `_rev` and `lastSequence` value. 
3. create a document locally (at client app). 
4. Push replicator replicates newly created document.
5. Push replicator calls `PUT /_local/{checkpointDocID}` -> Failed with 409
6. Push replicator calls `GET /_local/{checkpointDocID}` -> Receives latest checkpointDocID data.
7. Push replicator calls `PUT /_local/{checkpointDocID}` with local's `lastSequence` value

From network capture tool, only one 409 error by `PUT /_local/{checkpointDocID}`

![screen shot 2017-10-06 at 4 30 22 pm](https://user-images.githubusercontent.com/15889/31302138-2f34effa-aab4-11e7-9510-ce75ef7af781.png)

@NitzDKoder Did you observe observation item 3 with your environment?

**Comment by NitzDKoder** [comment]
@hideki please confirm what is the behaviour for pull continuous replication..our item 3 was with WRT to PULL replication..push seems all ok..

**Comment by hideki** [comment]
@NitzDKoder 

CBL v1.x behavior when `PUT /{db}/_local/{checkpointDocID}` has 409 error.

1. 409 error occurs because RevID for `checkpointDocID` is a mismatch.
2. CBL calls `GET /{db}/_local/{checkpointDocID}` to obtain SG's RevID and lastSequence values
3. CBL calls `PUT /{db}/_local/{checkpointDocID}` again with server's RevID and client's lastSequence values. This should not fail as RevID should match.

If you can observe Step 3 fails with 409, SG's `checkpointDocId`'s document is updated between Step 2 and 3.
